pkg/testgridanalysis/testgridconversion/to_raw_data_test.go

```diff
@@ -0,0 +1,574 @@
+				"container setup",
+			},
+			expectedTestNames: []string{
+				"[sig-cluster-lifecycle] Cluster completes upgrade",
+				"[sig-cluster-lifecycle] Cluster version operator acknowledges upgrade",
+				"[sig-mco] Machine config pools complete upgrade",
+				"[sig-sippy] upgrade should work",
+				"container setup",
+			},
+		},
+		// Overall has special meaning within the OpenShift Synthetic Test Manager
+		// as well as within the conversion code.
+		{
+			name: "Overall",
+			testGridTestNames: []string{
+				"Overall",
+			},
+			expectedTestNames: []string{"Overall"},
+			expectedWarnings: map[testgridv1.TestStatus][]string{
+				testgridv1.TestStatusFailure: missingSetupJobWarning,
+			},
+		},
+	}
+
+	options := testgridconversion.ProcessingOptions{
+		StartDay:             0,
+		NumDays:              numOfJobs,
+		SyntheticTestManager: testgridconversion.NewOpenshiftSyntheticTestManager(),
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			// Convert each test case into a RawDataTestCase and run it
+			runRawDataTestCase(t, rawDataTestCase{
+				testGridTestNames: testCase.testGridTestNames,
+				expectedTestNames: append(testCase.expectedTestNames, sippySyntheticTests...),
+				options:           options,
+				testFunc: func(opts testFuncOpts) {
+					// Ensure the warnings are what we expect
+					// Note: This func references the original test case, so must be defined within the test loop.
+					assertWarningsEqual(t, opts.warnings, testCase.expectedWarnings[opts.testGridStatus])
+				},
+			})
+		})
+	}
+}
+
+func TestToRawDataRunLengthEncoding(t *testing.T) {
+	// This test is intended to test the runlength encoding input from TestGrid
+	// as well as the numdays mechanism.
+	// Amongst other things, this test verifies that given a certain number of
+	// testgrid test statuses, that a specific rawtestresult will be generated.
+	testName := "test-1"
+
+	// We define a set of changelists which we will use across all test cases.
+	changelists := []string{
+		"123",
+		"456",
+		"789",
+		"987",
+		"654",
+		"321",
+		"ABC",
+		"DEF",
+	}
+
+	testCases := []struct {
+		// The name of the testcase for output purposes
+		name string
+		// Processing options for conversion code
+		options testgridconversion.ProcessingOptions
+		// The expected raw test result
+		expectedRawTestResult testgridanalysisapi.RawTestResult
+		// List of changelists which are expected to be failures.
+		failingChangelists []string
+		// List of changelists which are expected to be successes.
+		passingChangelists []string
+		// List of changelists which should not appear in the output given the
+		// startday / numday configuration.
+		excludedChangelists []string
+		// Note: Changelist lists within the test-case are unordered and no order is assumed.
+	}{
+		{
+			name: "minus one both ends",
+			options: testgridconversion.ProcessingOptions{
+				StartDay: -1,
```

Comment:
StartDay==-1 means "start at the first timestamp we find in the data", so I'd expect this test see the `123` changeset in the results, not sure why it doesn't?

Reply:
I'm also confused. I wonder if I didn't set up the test fixture correctly.

Reply:
I have an answer for this now:

If any of the timestamps is `time.Now()`, and `startDay` is 0, those timestamps are filtered out, because `computeLookback` also calls `time.Now()` and the resulting timestamp(s) fail these (https://github.com/openshift/sippy/blob/master/pkg/testgridanalysis/testgridconversion/to_raw_data.go#L79-L84) checks (they're not less than, they're equal to). For the sake of simplicity, the test fixtures subtract a second from `time.Now()` so they're not unintentionally filtered out.

Even with this being the case, we should still see `123` in the results since that's the first changelist and is aligned with the first timestamp. If `startDay` is set to `0`, the start column computed by `computeLookback` is also 0, hence `123` is present. However, if it is set to -1 (or any other negative number, for that matter) the start column gets set to the absolute value of `startDay` (effectively `n * -1`). This occurs regardless of whether the timestamp we're currently looking at occurs between `startDay` and `startDay + numDays`. Because of this, the first element in the timestamps list is ignored (which in this case, is `123`).

This is another case where the test reflects the current behavior, but I'm not sure if the current behavior is the desired behavoir.

Reply:
> However, if it is set to -1 (or any other negative number, for that matter) the start column gets set to the absolute value of startDay (effectively n * -1). 

the intended behavior is that if startDay is set to `-1`, then we effectively set startDay to align with(include) the latest timestamp in the dataset.  

Not sure where the absolute value bit is coming in, you can see that if startDay is -1, then this logic will end up with startTS==mostRecentTimestamp:
https://github.com/openshift/sippy/blob/9f8f55fd255647d7f9aacf698e78654e3279ea6c/pkg/testgridanalysis/testgridconversion/to_raw_data.go#L64-L73

I think the issue is that this logic:
https://github.com/openshift/sippy/blob/9f8f55fd255647d7f9aacf698e78654e3279ea6c/pkg/testgridanalysis/testgridconversion/to_raw_data.go#L79

is always going to exclude the actual column where the timestamp was the most recent timestamp(the one we found in the iteration above), because it's doing a less than check, not a less than or equal to check?

Reply:
The absolute value bit was just the best way to describe what was observed (the code isn't explicitly doing that, but that is the outcome). I didn't mean to imply the code was doing that explicitly.

The less-than-check is why the test fixtures decrement `time.Now()` by 1 second, since I was writing the tests to reflect the current behavior. What confused me was expecting `123` to be present when I couldn't see a way that the code (as-is) could do that. Since the intended behavior is to include those values, I'll change the less-than to less-than-or-equal. Please accept my apologies for overanalyzing.

Reply:
> Since the intended behavior is to include those values, I'll change the less-than to less-than-or-equal. 

thanks!  i'd say the original behavior was probably always bugged, it just didn't matter all that much if we missed one column, for the kind of reporting we do

> Please accept my apologies for overanalyzing.

no apologies necessary, it was a productive discussion.  And i'm happy to see some tests around some of this behavior... the code that extracts the "right" columns from testgrid based on the date range is definitely fiddly and subject to some off by one and other bounding issues (again, they don't matter that much in the grand scheme of what it needs to do, but tests that enforce the behavior are much appreciated)

Reply:
Duly noted, thanks! What you just described was my primary motivation for doing this work, especially since I want to use it as a baseline for my step metrics work. Is there anything else we should address at this point?

Reply:
i think this comment still needs resolution:
https://github.com/openshift/sippy/pull/243#discussion_r681154968

i think we should enforce the expectation that there is always an Overall test in the jobs we process, and then you can drop your testcase where there isn't one (or at least change the expected outcome of that test)